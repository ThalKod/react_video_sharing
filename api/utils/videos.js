const ffmpeg = require('fluent-ffmpeg');
const path = require("path");

const getDuration = (url, callback) =>{
  ffmpeg.ffprobe(url, (err, metadata) => {
    if(err) return callback({ error: true, msg: err});
    return callback({ error: false, duration: metadata.format.duration });
  });
};

const takeScreenshot = (video, duration, callback) => {
  const pathToScreenshot = path.join(__dirname, "../data", video.handle);
  let filename;
  const proc = ffmpeg(video.url)
      .on('filenames', (filenames) => {
        filename = filenames[0];
        console.log('filename : ' + filenames.join(', '));
      })
      .on('end', () => {
        console.log("Saved !");
        return callback({ error: false, pathToScreenshot: path.join(pathToScreenshot, filename) });
      })
      .on('error', (err) => {
        console.log("Error", err);
        return callback({ error: true, msg: err});
      })
      .takeScreenshots({ count: 1, timemarks: [ (duration / 4).toString() ], size: '1280x720' }, pathToScreenshot);
};

// process vidoe by taking a getting a video dutaion and a default screenscht
module.exports.processVideo = (video) => {
  return new Promise((resolve, reject) => {
    getDuration(video.url, (res) => {
      if(res.error) reject(res.msg);
      const duration = res.duration;
      takeScreenshot(video, duration, (result) => {
        if(result.error) reject(result.msg);
        const pathToScreenshot = result.pathToScreenshot;
        resolve({duration, pathToScreenshot});
      })
    })
  })
};
