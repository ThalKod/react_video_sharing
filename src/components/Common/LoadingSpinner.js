import React from "react";
import { BeatLoader } from "react-spinners";

export default ({text, medium}) => (
    <div className="content-wrapper">
      <div className="container">
        <div className="row">
          <div className={medium? "col-lg-6 upload-page": "col-lg-12 upload-page" }>
            <div className="u-area">
              <BeatLoader
                  sizeUnit="px"
                  size={30}
                  color='#123abc'
                  loading
              />
              {text && <p>{text}</p>}
            </div>
          </div>
        </div>
      </div>
    </div>
);
