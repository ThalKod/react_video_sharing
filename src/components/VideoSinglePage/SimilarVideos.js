import React from "react";

import { request } from "utils";
import VideoSingle from "components/HomePage/VideoSingle";

export default class SimilarVideos extends React.Component{

  state = {
    videos: [],
    loading: true,
  };

  componentDidMount = () => {
    const { id } = this.props;

    request("get", `/video/${id}/similar`)
        .then(res => {
          this.setState({ videos: res.data.videos, loading: false })
        })
        .catch(err => console.log(err));
  };

  renderVideos = () => {
    const { videos } = this.state;

    return videos.map((video) => {
      return <VideoSingle key={video._id} {...video}/>
    })
  };

  render(){
    const { loading, videos } = this.state;

    if(loading) return <div>Loading...</div>;

    if(videos.length <= 0) return null; // We don't show the similar videos section if we didn't find any

    return(
        <div>
          <div className="caption">
            <div className="left">
              <a href="/">Similar Videos</a>
            </div>
            <div className="clearfix"/>
          </div>
          <div className="list similar-videos">
            <div className="row">
              {this.renderVideos()}
            </div>
          </div>
        </div>
    )
  }
}
