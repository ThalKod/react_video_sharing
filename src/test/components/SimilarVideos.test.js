import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import SimilarVideos from "components/VideoSingle/SimilarVideos";
import LoadingSpinner from "components/Common/LoadingSpinner";
import videos from "test/__mock__/videos";

describe("SimilarVideos", () => {
  it("should render the Loading Spinner correctly while fetching videos", () => {
    const wrapper = shallow(<SimilarVideos/>);
    expect(
        wrapper.containsMatchingElement(
            <LoadingSpinner medium/>
        )
    ).toBe(true);
  });

  it("should render the videos list", () => {
    const wrapper = shallow(<SimilarVideos/>);
    wrapper.setState({ videos, loading: false });

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
