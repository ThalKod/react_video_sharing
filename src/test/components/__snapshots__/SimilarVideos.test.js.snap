// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimilarVideos should render the videos list 1`] = `
<div>
  <div
    className="caption"
  >
    <div
      className="left"
    >
      <a
        href="/"
      >
        Similar Videos
      </a>
    </div>
    <div
      className="clearfix"
    />
  </div>
  <div
    className="list similar-videos"
  >
    <div
      className="row"
    >
      <VideoSingle
        _id="wSLFcn28aaad3231"
        duration={13242}
        key="wSLFcn28aaad3231"
        name="My Video"
        viewCount={25000}
      />
      <VideoSingle
        _id="wSL32Fcn283231"
        duration={213242}
        key="wSL32Fcn283231"
        name="Test Video"
        viewCount={2500}
      />
      <VideoSingle
        _id="wSLFc930n283231"
        duration={213242}
        key="wSLFc930n283231"
        name="Video 2"
        viewCount={2500}
      />
      <VideoSingle
        _id="wSLFcn23s0983231"
        duration={2133242}
        key="wSLFcn23s0983231"
        name="New feature"
        viewCount={22500}
      />
      <VideoSingle
        _id="wSLFcn2829033231"
        duration={213242}
        key="wSLFcn2829033231"
        name="Video 23"
        viewCount={2500}
      />
    </div>
  </div>
</div>
`;
